docs/design-proposals/multiple-statefulsets.md

```diff
@@ -0,0 +1,201 @@
+* For on-premise deployment, k8s nodes may not have the same size. To make most of the resources, it's better to allow different profiles for a single cluster.
+* One TiKV peer use different resources(CPU/Memory/Storage) from other peers in the same TiDB cluster
+* No isolation between different tidb-servers, a big AP query may affect TP query. The worst case is that AP query cause a tidb-server OOM, and this tidb-server is also serving TP request
+* No support offline the specific peers
+* No support deploy TiDB Operator on multiple K8s clusters
+
+Using local persistent volume, once deployed the PD and TiKV pod has to be on the node all its lifecycle, migrating it to other nodes with larger resources would be hard. However if we use multiple statefulsets, we can do the migration on the fly. This is useful when there is a hot region on a TiKV that can hardly split.
+
+So the TiDB Operator should support creating multiple statefulsets for one component to achieve all these requests:
+
+* Add a new statefulset with bigger resources to scale TiDB cluster vertically
+* Different statefulsets can use different AZs or different resources and any other attributes
+* Can offline any peers when all the statefulsets’s replicas is 1
+* Deploy different statefulsets to different K8s clusters to support multiple K8s clusters(need further research)
+
+## Proposal
+
+### API change
+
+Users can create multiple statefulsets in `charts/tidb-cluster/values.yaml`, for example: change one tikv(Figure 1) to more tikvs(Figure 2):
+
+Figure 1:
+
+``` yaml
+tikv:
+  replicas: 3
+  image: pingcap/tikv:v2.1.0
+  …
+```
+
+Figure 2:
+
+``` yaml
+tikv:
+  replicas: 3
+  image: pingcap/tikv:v2.1.0
+  …
+  extra:
+  - name: name-1
+    replicas: 2
+    …
+  - name: name-2
+    replicas: 2
+    …
+```
+
+The default values.yaml is the same as of now. We can add document about how to extend the values.yaml to include the `extra` part listed here. But only a name is required, other fields can inherit from the upper level. And the generated `TidbCluster` object includes all the fields. And each statefulset has a separate ConfigMap named by the statefulset name, for example `name-1-tikv`. Also we can name the upper level statefulset as `default` so we can handle the names uniformly.
+
+We must add a name attribute(for example: name-1 or name-2) to the section to distinguish the different statefulsets. The name of the default statefulset is `default`.
+
+The same as TidbCluster object, from Figure 3 to Figure 4:
+
+Figure 3:
+
+``` go
+type TidbClusterSpec struct {
+  …
+  TiKV            TiKVSpec            `json:"tikv,omitempty"`
+  …
+}
+
+type TiKVStatus struct {
+StatefulSet     *apps.StatefulSetStatus     `json:"statefulSet,omitempty"`
+  …
+}
+```
+
+Figure 4:
+
+``` go
+type TidbClusterSpec struct {
+  TiKV            []TiKVSpec            `json:"tikv,omitempty"`
```

Comment:
`TiKV` is a unreasonable name when its type is an array.

Reply:
Any suggestion?

Reply:
@tennix @onlymellb @gregwebs Is this ok?

Reply:
I prefer `TiKVSets`

Reply:
`TiKVSets` is ok for me. Do we need rename `TiKVSpec` to `TiKVSetSpec`?